### PR TITLE
App Engine Service Split Traffic resource

### PR DESCRIPTION
**Pull request by khanali21**
* Added custom_create in custom_code.
* Added appengine service resource

# Release Note for Downstream PRs (will be copied)

```release-note:new-resource
`google_app_engine_service_split_traffic`
```

### CODE REVIEW

**Comment by khanali21** [comment]
@rileykarson I noticed that there is no option for the custom_create code in the generator. I thought it fitting to add this. This helps in Appengine Service Resource which does not have Create method.

If custom_create was not deliberately added, then I can move the resource under third_party as the custom resource.

Thanks
Ali

**Comment by khanali21** [comment]
Example template to follow.

**Comment by rileykarson** [comment]
For Terraform, I think we want to expose this under another name than `Service`. Since a service can't be (directly) created, I think it would be confusing to have a `google_app_engine_service` resource defined that depended on a version creating the service.


As a user, I'd expect my config to look like:

```hcl
resource "google_app_engine_service" "frontend" {
  id = "my-frontend-service"
  
} 

resource "google_app_engine_standard_app_version" "frontend_v2" {
  version_id = "v2"
  service = google_app_engine_service.frontend.id

  ...
}
```

When in reality, it's more likely to be like this, because the dependency is flipped:

```hcl
resource "google_app_engine_standard_app_version" "frontend_v2" {
  version_id = "v2"
  service = "my-frontend-service"

  ...
}

resource "google_app_engine_service" "frontend" {
  id = google_app_engine_standard_version.frontend_v2.service
  
} 
```

Instead, I'd lean towards exposing the resource as `google_app_engine_service_traffic_split`. I *believe* we can override the `name` value in `terraform.yaml` to accomplish that, but if not we'll need to add a duplicate entry in `api.yaml`.

**Comment by khanali21** [comment]
@rileykarson  Struggling with few things:
1. The create_verb can only be PUT and POST. Do you think I should modify ruby code to allow PATCH as the create_verb too.
2. The update_mask true includes the id field as well, however I want only one field "split" to be part of the updateMask, i.e. updateMask=split
3. I don't see the need for overriding name, as I changed the name of the resource and then also updated the resourceref for the Version resource. Let me know if there is any other need for name override. BTW I could not find how override the name, I did find legacy_name but could not make it work as intended. Can you tell me if that is the right one.


thanks

File: products/appengine/api.yaml [hunk 1]
```diff
@@ -73,24 +73,58 @@ objects:
...
         description: |
            Relative name of the service within the application. Example default.
+      - !ruby/object:Api::Type::NestedObject
+        name: 'split'
+        description: |
+           Mapping that defines fractional HTTP traffic diversion to different versions within the service.
+        required: false
+        properties:      
+          - !ruby/object:Api::Type::Enum
+            name: 'shardBy'
```
[the remark this answers is not in the document]
Comment (khanali21) [reply products/appengine/api.yaml]: I am going to test it shortly

File: products/appengine/terraform.yaml [hunk 2]
```diff
@@ -64,7 +64,16 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         test_env_vars:
           org_id: :ORG_ID
   Service: !ruby/object:Overrides::Terraform::ResourceOverride
-    exclude: true
+    id_format: "{{service_id}}"
+    import_format: ["apps/{{project}}/services/{{service_id}}"]
+    mutex: "apps/{{project}}/services/{{service_id}}"
+    custom_code: !ruby/object:Provider::Terraform::CustomCode
+      custom_create: templates/terraform/custom_create/noop_on_create_appengine_version.go.erb
```
[the remark this answers is not in the document]
Comment (khanali21) [reply products/appengine/terraform.yaml]: Not sure whether I need to add create_encoder in this case. So there are two things:
1. Having updateMask in the create_url, I think this can be handled by just adding the constant like:
```
    create_url: 'apps/{{project}}/services/{{service_id}}?{{migrateTraffic}}&updateMask=split'
````
2. Identifying if we need to add shardBy also in the updateMask (I am looking into it)

**Comment by khanali21** [comment]
@rileykarson  Kindly review this. I have made all the required changes, One question is still open.
Q. Should I just override the delete with no op as there is no deletion applicable in this resource.

**Comment by khanali21** [comment]
```
resource "google_storage_bucket" "bucket" {
	name = "<%= ctx[:vars]['bucket_name'] %>"
}

resource "google_storage_bucket_object" "object" {
	name   = "hello-world.zip"
	bucket = "${google_storage_bucket.bucket.name}"
	source = "./test-fixtures/appengine/hello-world.zip"
}

resource "google_app_engine_standard_app_version" "myapp_v1" {
  version_id = "v1"
  service = "myapp2"
  runtime = "nodejs10"
  noop_on_destroy = true
  entrypoint {
    shell = "node ./app.js"
  }
  deployment {
    zip {
      source_url = "https://storage.googleapis.com/${google_storage_bucket.bucket.name}/hello-world.zip"
    }  
  }
  env_variables = {
    port = "8080"
  } 

}
resource "google_app_engine_standard_app_version" "myapp_v2" {
  version_id = "v2"
  service = "myapp2"
  runtime = "nodejs10"
  noop_on_destroy = true
  entrypoint {
    shell = "node ./app.js"
  }
  deployment {
    zip {
      source_url = "https://storage.googleapis.com/${google_storage_bucket.bucket.name}/hello-world.zip"
    }  
  }
  env_variables = {
    port = "8080"
  } 

}

resource "google_app_engine_service_split_traffic" "myapp2" {
  service_id = "${google_app_engine_standard_app_version.myapp_v2.service}"
  migrate_traffic = false
  split {
    shard_by = "IP"
    allocations = {
      v1 = 0.75
      v2 = 0.25
    }
  }
}
```

File: products/appengine/api.yaml [hunk 3]
```diff
@@ -67,30 +67,74 @@ objects:
           - ALLOW
           - DENY
   - !ruby/object:Api::Resource
-    name: 'Service'
+    name: 'ServiceSplitTraffic'
```
Comment (rileykarson) [inline products/appengine/api.yaml]: Instead of renaming `Service`, would you mind duplicating this into a separate resource definition? It feels a little silly, but it's the best way to represent fine-grained resources right now.

File: products/appengine/api.yaml [hunk 4]
```diff
@@ -67,30 +67,74 @@ objects:
...
+          If set to true traffic will be migrated to this version.
     properties:
       - !ruby/object:Api::Type::String
         name: 'name'
-        output: true
         description: |
           Full path to the Service resource in the API. Example apps/myapp/services/default.
           This field is used in responses only. Any value specified here in a request is ignored.
       - !ruby/object:Api::Type::String
         name: 'id'
```
Comment (rileykarson) [inline products/appengine/api.yaml]: We have some liberty in terms of what we can call this field because it doesn't actually correspond to any fields we care about in the API. What do you think of `service` instead of `id`/`service_id`?

File: products/appengine/api.yaml (hunk 3 above)
Comment (rileykarson) [reply products/appengine/api.yaml]: This will also require exclusions in Ansible / InSpec for the Magician to work.

**Comment by khanali21** [comment]
@rileykarson Please review now.

Also here is my loud thinking on this:

Use case 1:
a) Enable appengine using google_app_engine_application resource
b) Create a new service named e.g. "myapp" and upload the code as v1 using google_app_engine_standard_app_version resource. (Please note that the service is automatically created if it is not existing). Use the noop_on_destroy to ensure that this version does not get deleted.
c) Create another version v2 for service myapp using google_app_engine_standard_app_version resource.
d) Split the traffic using the google_app_engine_service_split_traffic resource.

Problem: In this use case terraform destroy will leave the service myapp and version v1 when destroying all the resources.

Solution: Use the google_app_engine_service resource to delete the final version. For this we shall need the google_app_engine_service resource as I did in the first iteration which will ignore the create but allow deletion.

e) Destr

**Comment by khanali21** [comment]
@rileykarson Can you please review this. And give your comments.

File: products/appengine/api.yaml (hunk 4 above)
Comment (rileykarson) [reply products/appengine/api.yaml]: wdyt here?

File: products/appengine/api.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (rileykarson) [reply products/appengine/api.yaml]: Did you get a chance to test?

File: templates/terraform/examples/app_engine_service_split_traffic.tf.erb [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  depends_on = ["google_app_engine_standard_app_version.myapp_v1"]
+}
+
+resource "google_app_engine_service_split_traffic" "myapp" {
+  service_id = "${google_app_engine_standard_app_version.myapp_v2.service}"
+  migrate_traffic = false
+  split {
+    shard_by = "IP"
+    allocations = {
+      v1 = 0.75
```
Comment (rileykarson) [inline templates/terraform/examples/app_engine_service_split_traffic.tf.erb]: Are you able to interpolate the version names off of the `google_app_engine_standard_app_version` resource?

File: provider/terraform/custom_code.rb [hunk 6]
```diff
@@ -96,6 +96,10 @@ class CustomCode < Api::Object
       # method's function header can't be changed, the template
       # inserts that for you - do not include it in your custom code.
       attr_reader :custom_delete
+      # This code replaces the entire create method.  Since the create
```
Comment (rileykarson) [inline provider/terraform/custom_code.rb]: As long as we're not using this, do you mind removing it + the changes in `resource.erb`? I'd rather not add unused code.

**Comment by rileykarson** [comment]
Just a heads up & to be explicit, waiting on https://github.com/GoogleCloudPlatform/magic-modules/pull/2331 to get merged before making another pass here.

**Comment by khanali21** [comment]
@rileykarson that's right. 
A quick question. I am getting this error when running the test
				Resource specified by ResourceName couldn't be found: google_app_engine_service_split_traffic."

I believe this may be expected, so I thought I should put
				ExpectError:             regexp.MustCompile("Resource specified by ResourceName couldn't be found: google_app_engine_service_split_traffic."),

is that right approach?
Any pointer for a custom resource like split traffic.

File: third_party/terraform/tests/resource_app_engine_service_split_traffic_test.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample(context),
+			},
+			{
+				ResourceName:            "google_app_engine_service_split_traffic.",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"service_id", "split"},
+				ExpectError:             regexp.MustCompile("Resource specified by ResourceName couldn't be found: google_app_engine_service_split_traffic."),
```
Comment (rileykarson) [inline third_party/terraform/tests/resource_app_engine_service_split_traffic_test.go]: I don't believe this will get evaluated during an `ImportState` step, only a `Config` one.

Comment (khanali21) [reply third_party/terraform/tests/resource_app_engine_service_split_traffic_test.go]: yes, it does fail if I remove the line 35, which ExpectError. I did try to get to the bottom of this, but honestly could not understand. During debugging I found that it is failing after the test is finished, let me send you completed logs.

Comment (rileykarson) [reply third_party/terraform/tests/resource_app_engine_service_split_traffic_test.go]: Do you mind pushing this with the line removed? I'm curious why it's failing, and I'd like the test to be working before we merge this. No need for logs in that case, I can generate them myself.

Comment (khanali21) [reply third_party/terraform/tests/resource_app_engine_service_split_traffic_test.go]: Thanks for your review @rileykarson , please check this.

File: third_party/terraform/tests/resource_app_engine_service_split_traffic_test.go [hunk 8]
```diff
@@ -0,0 +1,127 @@
+package google
+
+import (
+	"log"
+	"regexp"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/acctest"
```
Comment (rileykarson) [inline third_party/terraform/tests/resource_app_engine_service_split_traffic_test.go]:
```suggestion
	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
```

Comment (rileykarson) [reply third_party/terraform/tests/resource_app_engine_service_split_traffic_test.go]: This + the other helper import need to be fixed to use the new plugin SDK package.

File: third_party/terraform/tests/resource_app_engine_service_split_traffic_test.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
```
Comment (rileykarson) [inline third_party/terraform/tests/resource_app_engine_service_split_traffic_test.go]: Is this import needed?

**Comment by khanali21** [comment]
@rileykarson I am unable to compile the code. I will try again today with clean setup.

**Comment by khanali21** [comment]
@rileykarson I am back working on this PR.  Can you do a quick review and let me know what it will take to get this merged. thanks in advance. 
I have rebased it on the latest master.

**Comment by khanali21** [comment]
@rileykarson So got over the issue with the compilation and also rebased it with the latest master. 
Will appreciate if we can get this merged as soon as possible.

**Comment by rileykarson** [comment]
Sorry for the delay in getting to this! I'm aware and have it on a list of things to do, we're preparing our `3.0.0` release and I've been short on time. I'm hoping to get to it either this Friday or mid-late next week.

**Comment by khanali21** [comment]
@rileykarson  did you get a chance to review it ?

File: products/appengine/terraform.yaml [hunk 10]
```diff
@@ -107,7 +107,31 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         primary_resource_id: "web_service"
         vars:
           bucket_name: "appengine-test-bucket"
-
+  ServiceSplitTraffic: !ruby/object:Overrides::Terraform::ResourceOverride
+    id_format: "{{service_id}}"
+    import_format: ["apps/{{project}}/services/{{service_id}}"]
+    mutex: "apps/{{project}}/services/{{service_id}}"
```
Comment (rileykarson) [inline products/appengine/terraform.yaml]: Most AppEngine operations need a mutex at the project level, `services/{{service_id}}` can be removed.

File: products/appengine/terraform.yaml [hunk 11]
```diff
@@ -107,7 +107,31 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         primary_resource_id: "web_service"
         vars:
           bucket_name: "appengine-test-bucket"
-
+  ServiceSplitTraffic: !ruby/object:Overrides::Terraform::ResourceOverride
+    id_format: "{{service_id}}"
```
Comment (rileykarson) [inline products/appengine/terraform.yaml]: `id_format` and `import_format` are probably redundant with what's used by default. Did you find that you needed custom values?

File: products/appengine/terraform.yaml [hunk 12]
```diff
@@ -107,7 +107,31 @@ overrides: !ruby/object:Overrides::ResourceOverrides
...
+    examples:
+      - !ruby/object:Provider::Terraform::Examples
+        name: "app_engine_service_split_traffic"
+        vars:
+          service_id: "default"
+          split.allocations.v1: "1"
+          bucket_name: "appengine-static-content"
+        test_env_vars:
+          org_id: :ORG_ID    
+        skip_test: true
```
Comment (rileykarson) [inline products/appengine/terraform.yaml]: Do you remember why we had to skip adding a test? I think we can generate one.

File: products/appengine/api.yaml [hunk 13]
```diff
@@ -502,4 +502,74 @@ objects:
...
+      - !ruby/object:Api::Type::Boolean
+        name: 'migrate_traffic'
+        url_param_only: true
+        description: |
+          If set to true traffic will be migrated to this version.
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'name'
+        description: |
+          Full path to the Service resource in the API. Example apps/myapp/services/default.
```
Comment (rileykarson) [inline products/appengine/api.yaml]: "The name of the service these settings apply to."

File: products/appengine/api.yaml [hunk 14]
```diff
@@ -502,4 +502,74 @@ objects:
                 Pathname within the host. Must start with a "/". A single "*" can be included at the end of the path.
                 The sum of the lengths of the domain and path may not exceed 100 characters.
               required: true
+  - !ruby/object:Api::Resource
+    name: 'ServiceSplitTraffic'
+    description: |
+      A Service resource is a logical component of an application that can share state and communicate in a secure fashion with other services.
```
Comment (rileykarson) [inline products/appengine/api.yaml]: "Traffic routing configuration for versions within a single service. Traffic splits define how traffic directed to the service is assigned to versions."

File: products/appengine/api.yaml [hunk 15]
```diff
@@ -502,4 +502,74 @@ objects:
...
+        url_param_only: true
+        description: |
+          If set to true traffic will be migrated to this version.
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'name'
+        description: |
+          Full path to the Service resource in the API. Example apps/myapp/services/default.
+          This field is used in responses only. Any value specified here in a request is ignored.
+      - !ruby/object:Api::Type::String
```
Comment (rileykarson) [inline products/appengine/api.yaml]: This seems redundant with `name`/`service`

File: products/appengine/api.yaml [hunk 16]
```diff
@@ -502,4 +502,74 @@ objects:
...
+        message: 'message'
+    parameters:
+      - !ruby/object:Api::Type::Boolean
+        name: 'migrate_traffic'
+        url_param_only: true
+        description: |
+          If set to true traffic will be migrated to this version.
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'name'
```
Comment (rileykarson) [inline products/appengine/api.yaml]:
```suggestion
        name: 'service'
```

File: products/appengine/api.yaml [hunk 17]
```diff
@@ -502,4 +502,74 @@ objects:
...
+           Relative name of the service within the application. Example default.
+      - !ruby/object:Api::Type::NestedObject
+        name: 'split'
+        description: |
+           Mapping that defines fractional HTTP traffic diversion to different versions within the service.
+        required: true
+        properties:      
+          - !ruby/object:Api::Type::Enum
+            name: 'shardBy'
+            required: false
```
Comment (rileykarson) [inline products/appengine/api.yaml]:
```suggestion
```

If not required, you can omit `required` entirely.

File: templates/terraform/examples/app_engine_service_split_traffic.tf.erb (hunk 5 above)
Comment (rileykarson) [reply templates/terraform/examples/app_engine_service_split_traffic.tf.erb]: This is unaddressed.

File: third_party/terraform/tests/resource_app_engine_service_split_traffic_test.go (hunk 7 above)
Comment (rileykarson) [reply third_party/terraform/tests/resource_app_engine_service_split_traffic_test.go]: Can you remove this? I'd like to get this working (and ideally, generate the test)

File: products/appengine/terraform.yaml (hunk 12 above)
Comment (khanali21) [reply products/appengine/terraform.yaml]: That's the problem, if I do not skip_test and generate it I get the error which I fixed using the work around. I have done quite a bit of debugging but could not reach to the bottom of this issue.

Comment (rileykarson) [reply products/appengine/terraform.yaml]: Would you mind pushing a commit with the test enabled so I can take a look again? I'd really like to get it working.

**Comment by khanali21** [comment]
I removed the skip_test and also removed the template for the test.
So now the test is being auto-generated.
I get this error:
```
root@84391ff50932:/mm# ./integration_tests.sh
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
=== PAUSE TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
=== CONT  TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
--- FAIL: TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample (128.60s)
    testing.go:635: Step 1 error: Resource specified by ResourceName couldn't be found: google_app_engine_service_split_traffic.
FAIL
FAIL    github.com/terraform-providers/terraform-provider-google/google 128.629s
GNUmakefile:19: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

**Comment by khanali21** [comment]
> I removed the skip_test and also removed the template for the test.
> So now the test is being auto-generated.
> I get this error:
> 
> ```
> root@84391ff50932:/mm# ./integration_tests.sh
> ==> Checking source code against gofmt...
> ==> Checking that code complies with gofmt requirements...
> TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
> === RUN   TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
> === PAUSE TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
> === CONT  TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
> --- FAIL: TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample (128.60s)
>     testing.go:635: Step 1 error: Resource specified by ResourceName couldn't be found: google_app_engine_service_split_traffic.
> FAIL
> FAIL    github.com/terraform-providers/terraform-provider-google/google 128.629s
> GNUmakefile:19: recipe for target 'testacc' failed
> make: *** [testacc] Error 1
> ```

I had debugged this and somehow could not understand the issue. The work around I found was to add the expected error as the regex.

File: products/appengine/api.yaml [hunk 18]
```diff
@@ -515,4 +515,70 @@ objects:
...
+    base_url: 'apps/{{project}}/services'
+    create_url: 'apps/{{project}}/services/{{service_id}}?migrateTraffic={{migrate_traffic}}&updateMask=split'
+    create_verb: :PATCH
+    update_url: 'apps/{{project}}/services/{{service_id}}?migrateTraffic={{migrate_traffic}}'
+    update_verb: :PATCH
+    update_mask: true
+    references: !ruby/object:Api::Resource::ReferenceLinks
+      api: 'https://cloud.google.com/appengine/docs/admin-api/reference/rest/v1/apps.services'
+    async: !ruby/object:Api::Async
+      operation: !ruby/object:Api::Async::Operation
```
Comment (rileykarson) [inline products/appengine/api.yaml]:
```suggestion
      operation: !ruby/object:Api::OpAsync::Operation
```

`Async` was renamed `OpAsync`, do you mind updating that? Sorry for the delay in getting back!

Comment (khanali21) [reply products/appengine/api.yaml]: Thanks @rileykarson.

File: products/appengine/api.yaml [hunk 19]
```diff
@@ -515,4 +515,70 @@ objects:
...
+      Traffic routing configuration for versions within a single service. Traffic splits define how traffic directed to the service is assigned to versions.
+    base_url: 'apps/{{project}}/services'
+    create_url: 'apps/{{project}}/services/{{service_id}}?migrateTraffic={{migrate_traffic}}&updateMask=split'
+    create_verb: :PATCH
+    update_url: 'apps/{{project}}/services/{{service_id}}?migrateTraffic={{migrate_traffic}}'
+    update_verb: :PATCH
+    update_mask: true
+    references: !ruby/object:Api::Resource::ReferenceLinks
+      api: 'https://cloud.google.com/appengine/docs/admin-api/reference/rest/v1/apps.services'
+    async: !ruby/object:Api::Async
```
Comment (rileykarson) [inline products/appengine/api.yaml]: I think all the occurrences of `::Async` here need to be `::OpAsync`, but I'm not sure. Can you amend them and make sure that it compiles locally?

Comment (khanali21) [reply products/appengine/api.yaml]: hmm... I was able to compile. But let me check after rebase.

Comment (khanali21) [reply products/appengine/api.yaml]: @rileykarson I went on long vacation, came back.
I have checked in the latest code with the rebase on the current master.
Everything fine but I am still getting the following error:
root@52d8e96b9325:/mm# ./integration_tests.sh
==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
=== PAUSE TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
=== CONT  TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample
--- FAIL: TestAccAppEngineServiceSplitTraffic_appEngineServiceSplitTrafficExample (136.23s)
    testing.go:635: Step 1 error: Resource specified by ResourceName couldn't be found: google_app_engine_service_split_traffic.
FAIL
FAIL    github.com/terraform-providers/terraform-provider-google/google 136.256s
GNUmakefile:19: recipe for target 'testacc' failed
make: *** [testacc] Error 1

File: products/appengine/terraform.yaml [hunk 20]
```diff
@@ -109,7 +109,30 @@ overrides: !ruby/object:Overrides::ResourceOverrides
...
+        name: 'service_id'
+        ignore_read: true  
+      split: !ruby/object:Overrides::Terraform::PropertyOverride
+        ignore_read: true    
+      split.allocations: !ruby/object:Overrides::Terraform::PropertyOverride
+        custom_flatten: templates/terraform/custom_flatten/float64_to_string.go.erb
+    examples:
+      - !ruby/object:Provider::Terraform::Examples
+        name: "app_engine_service_split_traffic"
+        vars:
```
Comment (rileykarson) [inline products/appengine/terraform.yaml]:
```suggestion
        primary_resource_id: 'myapp'
        vars:
```

This should resolve the issue you were seeing. It was caused when the test framework effectively ran `terraform import google_app_engine_service_split_traffic. {{id}}` instead of `terraform import google_app_engine_service_split_traffic.myapp {{id}}`

File: products/appengine/api.yaml (hunk 19 above)
Comment (rileykarson) [reply products/appengine/api.yaml]: Replied in a comment below.

File: products/appengine/terraform.yaml (hunk 20 above)
Comment (khanali21) [reply products/appengine/terraform.yaml]: Cool, thanks I have committed the suggestion.

File: products/appengine/terraform.yaml (hunk 11 above)
Comment (rileykarson) [reply products/appengine/terraform.yaml]: Resolving since this is really old.

File: products/appengine/terraform.yaml [hunk 21]
```diff
@@ -109,7 +109,31 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         primary_resource_id: "web_service"
         vars:
           bucket_name: "appengine-test-bucket"
-
+  ServiceSplitTraffic: !ruby/object:Overrides::Terraform::ResourceOverride
+    id_format: "{{service_id}}"
```
Comment (rileykarson) [inline products/appengine/terraform.yaml]: Can you try removing the `id_format` and `import_format`? If they need to be there, can you comment with what the format is when they're not set?

We try to make `id` values look like `apps/{{project}}/services/{{service_id}}` now, so that should be the default behaviour. If the generated ones are wrong, you can use that form for both `id` and `import` I think.

File: templates/terraform/examples/app_engine_service_split_traffic.tf.erb [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+      v2 = 0.25
+    }
+  }
+  depends_on = ["google_app_engine_standard_app_version.myapp_v2"]
+}
+resource "google_app_engine_service_split_traffic" "myapp2" {
```
Comment (rileykarson) [inline templates/terraform/examples/app_engine_service_split_traffic.tf.erb]: I think we can remove this entry, having just the first is fine.

File: templates/terraform/examples/app_engine_service_split_traffic.tf.erb (hunk 5 above)
Comment (rileykarson) [reply templates/terraform/examples/app_engine_service_split_traffic.tf.erb]: Resolving as old.

File: templates/terraform/examples/app_engine_service_split_traffic.tf.erb [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+  service_id = "${google_app_engine_standard_app_version.myapp_v2.service}"
+  migrate_traffic = false
+  split {
+    shard_by = "IP"
+    allocations = {
+      v1 = 0.75
+      v2 = 0.25
+    }
+  }
+  depends_on = ["google_app_engine_standard_app_version.myapp_v2"]
```
Comment (rileykarson) [inline templates/terraform/examples/app_engine_service_split_traffic.tf.erb]: This is redundant with the interpolation in `service_id` I think, and can be removed.

Comment (khanali21) [reply templates/terraform/examples/app_engine_service_split_traffic.tf.erb]: This is needed in order to get the integration tests passed; as the test fail because it can not destroy the service which has zero traffic.

File: products/appengine/terraform.yaml (hunk 21 above)
Comment (khanali21) [reply products/appengine/terraform.yaml]: I tried. But it fails if I try to remove the the id_format and import_format , or change the id_format to apps/{{project}}/services/{{service_id}}. I am doing more testing to find out the reason for that.

Comment (rileykarson) [reply products/appengine/terraform.yaml]: Can post the error you get when setting `id_format` to `['apps/{{project}}/services/{{service_id}}']`? Using these values is fine if the defaults aren't working.

File: templates/terraform/examples/app_engine_service_split_traffic.tf.erb (hunk 23 above)
Comment (rileykarson) [reply templates/terraform/examples/app_engine_service_split_traffic.tf.erb]: I think the same connection between resources is established by https://github.com/GoogleCloudPlatform/magic-modules/pull/2269/files#diff-458f601ca7d7a40a89cb7b8bf8595b19R49, though?

Comment (khanali21) [reply templates/terraform/examples/app_engine_service_split_traffic.tf.erb]: @rileykarson 
So I tested it , here is what is happening. We create v1 it takes 100% of the traffic, we create v2 it has no traffic, we use split traffic resource to create 75-25% split between the two versions.
Now when the destroy happens, it tries to destroy v1 which is having 75% traffic, and it fails as the v1 is part of split traffic, I don't think it automatically transfers the traffic to other v1.

What happens when I add the second split, it migrate all the traffic to v2, then destroy works fine.

File: products/appengine/terraform.yaml (hunk 21 above)
Comment (khanali21) [reply products/appengine/terraform.yaml]: @rileykarson 
It is fixed. Don't get any error with id_format as "apps/{{project}}/services/{{service_id}}".
It was my bad.

File: templates/terraform/resource.erb [hunk 24]
```diff
@@ -131,9 +131,6 @@ func resource<%= resource_name -%><%= prop.name.camelize(:upper) -%>SetStyleDiff
 <% end -%>
 
 func resource<%= resource_name -%>Create(d *schema.ResourceData, meta interface{}) error {
-<%  if object.custom_code.custom_create -%>
```
Comment (rileykarson) [inline templates/terraform/resource.erb]: I think I said to remove this, but it looks already committed. Never mind in that case- can you revert the changes to this file?

**Comment by khanali21** [comment]
@rileykarson Apologies for my complacency , I am starting getting the following error after applying the changes
==> Checking that code complies with gofmt requirements...
go install
# github.com/terraform-providers/terraform-provider-google/google
google/error_retry_predicates.go:62:23: urlerr.Unwrap undefined (type *url.Error has no field or method Unwrap)
google/error_retry_predicates.go:76:24: urlerr.Unwrap undefined (type *url.Error has no field or method Unwrap)
GNUmakefile:11: recipe for target 'build' failed
make: *** [build] Error 2

I am trying to dig it.

**Comment by khanali21** [comment]
@rileykarson I am struggling with the following error, apparently this is due to a change 
Do I need to update something?


commit c94b6c7206d4bba55909c0ed84b83560a7f4bbc3
Author: The Magician <magic-modules@google.com>
Date:   Thu Feb 13 12:29:36 2020 -0800

    Add more generic error predicates (#3116) (#5675)

    * add connection reset by peer predicate

    * lint

    * separate error preds, fix temp pred

    * Fix error retries, comments

    * remove removed func

    Signed-off-by: Modular Magician <magic-modules@google.com>




 
==> Checking that code complies with gofmt requirements...
go install
# github.com/terraform-providers/terraform-provider-google/google
google/error_retry_predicates.go:62:23: urlerr.Unwrap undefined (type *url.Error has no field or method Unwrap)
google/error_retry_predicates.go:76:24: urlerr.Unwrap undefined (type *url.Error has no field or method Unwrap)
GNUmakefile:11: recipe for target 'build' failed
make: *** [build] Error 2

**Comment by khanali21** [comment]
@rileykarson Still struggling with it, I have tried even with the old code that was passing the compilation. I have also tried reverting the changes in module terraform provider.
Any help is appreciated. Thanks in advance.

File: products/appengine/api.yaml [hunk 25]
```diff
@@ -515,4 +515,69 @@ objects:
...
+    parameters:
+      - !ruby/object:Api::Type::Boolean
+        name: 'migrate_traffic'
+        url_param_only: true
+        description: |
+          If set to true traffic will be migrated to this version.
+    properties:
+      - !ruby/object:Api::Type::String
+        name: 'service'
+        url_param_only: true
```
Comment (rileykarson) [inline products/appengine/api.yaml]:
```suggestion
        api_name: 'id'
```

File: templates/terraform/resource.erb (hunk 24 above)
Comment (rileykarson) [reply templates/terraform/resource.erb]: These changes still need to be removed.

**Comment by khanali21** [comment]
@rileykarson DONE :) Had a bad time with this one.... even the changes from resource.erb were not going away git acting up with resolving merge conflict.
